### PR TITLE
Feature/OPDLG 330  Dynamic attributes

### PR DESCRIPTION
**Pull request by bendeeson**
Adds the ability to create 'dynamic attributes' that are stored in the database rather than config files.

Depends on this PR for core: https://github.com/opendialogai/core/pull/428

Adds new `/admin/api/dynamic-attribute` api resource with support for CRUD of DynamicAttribute model. 
Adds new `/admin/api/dynamic-attributes/{upload,download}` (note the s) routes for uploading/downloading all dynamic-attributes as JSON. 

Adds new `dynamic-attributes:import` and `dynamic-attributes:export` artisan commands for exporting the dynamic-attributes to a .json file.

See app/Http/Controllers/API/DynamicAttributesController.php

### CODE REVIEW

File: routes/api.php [hunk 1]
```diff
@@ -17,47 +17,96 @@
...
+        );
 
-    Route::get('requests', 'RequestsController@index');
-    Route::get('requests/{id}', 'RequestsController@show');
+        Route::apiResource('global-context', 'GlobalContextsController');
 
-    Route::get('warnings', 'WarningsController@index');
-    Route::get('warnings/{id}', 'WarningsController@show');
+        
+        Route::get('dynamic-attributes/download', 'DynamicAttributesController@download');
```
Comment (bendeeson) [inline routes/api.php:54]: Note that this route is `dynamic-attributes` with an s. Avoids overloading the `dynamic-attribute/download` route with two meanings (download Dynamic-attributes/Get the dynamic attribute with id 'download')

File: app/Http/Controllers/API/DynamicAttributesController.php [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+    public function download(): JsonResponse
+    {
+        $map = [];
+        foreach (DynamicAttribute::all() as $dynamicAttribute) {
+            $map[$dynamicAttribute->id] = $dynamicAttribute->type;
+        }
+        return new JsonResponse(
+            $map,
+            200,
+            ['Content-Disposition' => 'attachment; filename="dynamic-attributes.json"']
```
Comment (elliotmassen) [inline app/Http/Controllers/API/DynamicAttributesController.php:226]: I'm thinking the filename should be `custom-attributes.json`.

**Review by elliotmassen — CHANGES_REQUESTED** [review]
This looks great! 🔥  Just a few comments and the following:

 - [ ] Would you be able to commit the Composer file updates to use the related `opendialogai/core` dependency? Ie. to update it to be `dev-feature/OPNDLG-330--dynamic-attributes-crud`. We'll need to change it again before merging, but just keeps it related and makes it easier to pull and test before approving.
 - [ ] Could you add some endpoint tests similar to what we do in `ConversationsTest`?

File: composer.json [hunk 3]
```diff
@@ -12,7 +12,7 @@
         "laravel/framework": "^6.0",
         "laravel/tinker": "^1.0",
         "maennchen/zipstream-php": "^2.0",
-        "opendialogai/core": "0.7.x-dev",
+        "opendialogai/core": "dev-feature/OPNDLG-330--dynamic-attributes-crud",
```
Comment (elliotmassen) [inline composer.json:15]: Just leaving this comment to remind us to merge the core changes prior to this work and to update this depedency to the latest `1.x` afterwards.

**Review by elliotmassen — APPROVED** [review]

**Comment by elliotmassen** [comment]
Thanks for making the changes - the tests look great :)

**Comment by elliotmassen** [comment]
Closing in favour of #237.

**Pull request closed without merging**